### PR TITLE
Fix biometrics on android

### PR DESCRIPTION
**Pull request by brunobar79**
Fixes RAI-865

- Added face unlock support for android
- Added fingerprint support for android
- Added custom PIN support for android if no biometrics detected
- Improved the "Portal" usage to have only one instance at the top level of the app,
that way any screen can show / dismiss the current loading overlay

SCENARIOS THAT NEED TO BE COVERED on both OS:

1 - New wallet - Basic test flow with face auth enabled
2 - Importing wallet - Basic + TX flow with with face auth enabled
3 - New wallet - Basic test flow with fingerprint enabled
4 - Importing wallet - Basic + TX flow with fingerprint enabled
5 - New wallet - Basic test flow with device PIN enabled (should fallback to custom PIN)
6 - Importing wallet - Basic + TX flow with device PIN enabled (should fallback to custom PIN)
7 - Restore wallet - Basic + TX flow with face auth enabled
8- Restore wallet - Basic + TX flow with fingerprint enabled
9 - Restore wallet - Basic + TX flow with device PIN enabled (should fallback to custom PIN)

Basic test flow:
- Launch the app from clean state
- Try backups & check seed

TX flow:
- Send some asset from the send flow
- Swap
- WC

(This can be done with ganache. Mainnet is not required)

WHAT"S NOT COVERED BY THIS PR:

1 - Upgrading security:  If you have custom PIN and then you upgrade to fingerprint or face auth you'll get an error that says "Authentication error"

We can handle in a separate PR, where we can keep track of the auth method used while storing the key and then always use that one. If a user wants to change it they can backup and reimport.

### CODE REVIEW

**Comment by osdnk** [comment]
Sorry I haven't managed to take a look before. I'll try to do it asap

**Comment by osdnk** [comment]
I've sent comments in PM on Slack!

**Comment by skylarbarrera** [comment]
All Android flows look good, the following errors I experienced on a variety depending on which type of auth was used. Still going through iOS but so far no regressions. 

Hold to Send Button pretty flaky, faceID & touchID: https://cloud.skylarbarrera.com/Screen_Recording_20201021-161107.mp4

FaceID back up loop(worked the first time, got this loop 3 times after): https://cloud.skylarbarrera.com/Screen_Recording_20201021-165824.mp4

Hangs for a moment after authorizing transaction(all 3 but intermittent): https://cloud.skylarbarrera.com/Screen_Recording_20201021-160911.mp4 

Will go through again after these are settled!

**Comment by osdnk** [comment]
I see one problem. Let’s say I want to say 1234 as a pin. Then I type 124 by accident. I can easily revert 4 and press 3. However, when I inputted 1235, I cannot return. Usually, there’s a way to go back from the PIN confirmation state

**Comment by brunobar79** [comment]
@osdnk  Just pushed a commit that handles the scenario you mentioned on your last comment, as we discussed on Slack

**Comment by osdnk** [comment]
It's good. let's merge it

**Review by osdnk — APPROVED** [review]

File: src/components/buttons/hold-to-authorize/HoldToAuthorizeButton.js [hunk 1]
```diff
@@ -74,6 +75,14 @@ const calculateReverseDuration = proc(longPressProgress =>
   multiply(divide(longPressProgress, 100), longPressProgressDurationMs)
 );
 
+const LoadingSpinner = styled(android ? Spinner : ActivityIndicator).attrs({
+  color: 'white',
```
Comment (osdnk) [inline src/components/buttons/hold-to-authorize/HoldToAuthorizeButton.js:79]:
```suggestion
  color: colors.white,
```


+ import

File: src/model/wallet.ts [hunk 2]
```diff
@@ -1002,6 +1159,25 @@ export const loadSeedPhraseAndMigrateIfNeeded = async (
...
       seedPhrase = get(seedData, 'seedphrase', null);
+      let userPIN = null;
+      if (android) {
+        const hasBiometricsEnabled = await getSupportedBiometryType();
+        // Fallback to custom PIN
+        if (!hasBiometricsEnabled) {
+          try {
+            userPIN = await authenticateWithPIN();
+            if (userPIN) {
+              // Encrypt with the PIN
```
Comment (osdnk) [inline src/model/wallet.ts:1170]:
```suggestion
              // Dencrypt with the PIN
```

**Comment by skylarbarrera** [comment]
Samsung FaceID(on) touchID (off), i.e only faceID enabled, needs to revert to set/enter pin instead of 'try different biometric alert' : https://cloud.skylarbarrera.com/samsung-faceID-fallback.mp4

If theres a way to tell without asking for biometrics, maybe just add a check on new/import?

File: src/screens/ImportSeedPhraseSheet.js [hunk 3]
```diff
@@ -361,16 +351,11 @@ export default function ImportSeedPhraseSheet() {
...
     if (isImporting) {
-      setComponent(
-        <LoadingOverlay
-          paddingTop={keyboardVerticalOffset}
-          title={walletLoadingStates.IMPORTING_WALLET}
-        />,
-        true
-      );
-      return hide;
+      setIsWalletLoading(walletLoadingStates.IMPORTING_WALLET);
```
Comment (jinchung) [inline src/screens/ImportSeedPhraseSheet.js:354]: minor: can we replace the if else with something like `setIsWalletLoading(isImporting ? walletLoadingStates.IMPORTING_WALLET : null);`

File: src/helpers/biometryTypes.js [hunk 4]
```diff
@@ -1,5 +1,7 @@
 export default {
+  Face: 'Face', // Android
   FaceID: 'FaceID',
+  Fingerprint: 'Fingerprint',
```
Comment (jinchung) [inline src/helpers/biometryTypes.js:4]: Did you care about also noting "Fingerprint" as // Android

File: src/model/wallet.ts [hunk 5]
```diff
@@ -930,14 +1056,39 @@ const migrateSecrets = async (): Promise<MigratedSecretsResult | null> => {
       return null;
     }
 
+    let userPIN = null;
+    if (android) {
```
Comment (jinchung) [inline src/model/wallet.ts:1060]: Android users should never be able to reach this code given that they should never need to migrateSecrets

File: src/screens/PinAuthenticationScreen.js [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+const PinAuthenticationScreen = () => {
+  useBlockBackButton();
+  const { params } = useRoute();
+  const { goBack, setParams } = useNavigation();
+  const [errorAnimation, onShake] = useShakeAnimation();
+
+  const { isNarrowPhone, isSmallPhone, isTallPhone } = useDimensions();
+
+  const [attemptsLeft, setAttemptsLeft] = useState(MAX_ATTEMPTS_LEFT);
```
Comment (jinchung) [inline src/screens/PinAuthenticationScreen.js:37]: If a user kills the app before they reach the MAX_ATTEMPTS_LEFT, does this mean they could have unlimited retries?

File: src/screens/PinAuthenticationScreen.js [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+    };
+  }, [params, setParams]);
+
+  useEffect(() => {
+    const checkTimelock = async () => {
+      // When opening the screen we need to check
+      // if the user wasn't banned for too many tries
+      const timelock = await getAuthTimelock();
+      if (timelock) {
+        const stillBanned = Date.now() < timelock;
```
Comment (jinchung) [inline src/screens/PinAuthenticationScreen.js:60]: Can reuse Date.now() (on line 60 and 62) and can set timeLeftMS to timelock - dateNow (instead of dateNow - timelock) so you don't need the Math.abs later

File: src/model/wallet.ts [hunk 8]
```diff
@@ -414,11 +422,28 @@ const loadPrivateKey = async (): Promise<null | EthereumPrivateKey | -1> => {
...
+
       const privateKeyData = await getPrivateKey(address);
       if (privateKeyData === -1) {
         return -1;
       }
       privateKey = get(privateKeyData, 'privateKey', null);
+
+      let userPIN = null;
+      if (android) {
+        const hasBiometricsEnabled = await getSupportedBiometryType();
```
Comment (jinchung) [inline src/model/wallet.ts:434]: Not urgent / we can do this in a separate PR, but some of these PIN checks are very similar to each other in multiple places in this file, and would be nice to have a separate function if it makes sense to do so

Comment (brunobar79) [reply src/model/wallet.ts:427]: I agree this can be improved but I don't wanna touch it now cause it has been QA'ed already and it's a lot of cases.
I'll keep this in mind if we introduce PIN for iOS or lockscreen.

**Review by jinchung — APPROVED** [review]
🌮

File: src/model/wallet.ts [hunk 9]
```diff
@@ -729,6 +802,15 @@ export const getPrivateKey = async (
     const pkey = (await keychain.loadObject(key, {
       authenticationPrompt,
     })) as PrivateKeyData;
+
+    if (pkey === -2) {
+      Alert.alert(
+        'Error',
+        'Your current authentication method (Face Recognition) is not secure enough, please go to "Settings > Biometrics & Security" and enable an alternative biometric method like Fingerprint or Iris'
```
Comment (skylarbarrera) [inline src/model/wallet.ts:809]:
```suggestion
        'Your current authentication method (Face Recognition) is not secure enough, please go to "Settings > Biometrics & Security" and enable an alternative biometric method like Fingerprint or Iris.'
```

**Review by skylarbarrera — APPROVED** [review]
looks good, ran through the whole gambit, no issues! 🥳
